File: src/main/java/com/bovilexics/javaph/qi/QiServerFileReader.java
package com.bovilexics.javaph.qi;

import com.bovilexics.javaph.logging.ErrLogger;
import org.jetbrains.annotations.NotNull;

import java.io.FileNotFoundException;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public final class QiServerFileReader
{
    private static final String SEPARATOR = "::";
    private FieldFactory factory = null;

    public QiServerFileReader(final @NotNull FieldFactory factory)
    {
        this.factory = factory;
    }

    public @NotNull List<Server> readServers(final @NotNull String fileName) throws FileNotFoundException, IOException
    {
        final @NotNull Path path = Paths.get(fileName);
        final @NotNull List<String> lines = Files.readAllLines(path);
        final @NotNull List<Server> results = new ArrayList<>(lines.size());

        for (int i = 0; i < lines.size(); i++)
        {
            final @NotNull String line = lines.get(i);
            // Ignore comment lines
            if (line.startsWith("#")) {
                continue;
            }

            final @NotNull String[] items = line.split(SEPARATOR);

            if (items.length != 3)
            {
                ErrLogger.instance.println("Error: Invalid server entry in " + fileName + " on line " + i + " --> " + line);
            }
            else
                {
                final @NotNull Server server = new QiServer(factory, items[0], items[1], items[2]);
                results.add(server);
            }
        }
        return Collections.unmodifiableList(results);
    }
}

File: src/main/scala/com/bovilexics/javaph/qi/QiServerFileReader.scala
package com.bovilexics.javaph.qi

import java.io.{FileNotFoundException, IOException}
import java.nio.file.{Files, Paths}
import java.util
import java.util.Collections

import com.bovilexics.javaph.logging.ErrLogger

object QiServerFileReader
{
  val SEPARATOR = "::"
}

final class QiServerFileReader(val serverFactory: ServerFactory)
{
  @throws[FileNotFoundException]
  @throws[IOException]
  def readServers(fileName: String): util.List[Server] =
  {
    val path = Paths.get(fileName)
    val lines = Files.readAllLines(path)
    val results = new util.ArrayList[Server](lines.size)
    val separator = QiServerFileReader.SEPARATOR
    for
    {
      i <- 0 to lines.size()
      if !lines.get(i).startsWith("#") // Ignore comment lines
    }
    {
      val line = lines.get(i)
      val items = line.split(separator)
      if (items.length != 3)
      {
        ErrLogger.instance.println("Error: Invalid server entry in " + fileName + " on line " + i + " --> " + line)
      }
      else
      {
        val server = serverFactory.create(items(0), items(1), items(2))
        results.add(server)
      }
    }
    Collections.unmodifiableList(results)
  }
}